src/org/mockito/internal/configuration/injection/PropertyAndSetterInjection.java

```diff
@@ -125,27 +124,45 @@ private boolean injectMockCandidatesOnFields(Set<Object> mocks, Object instance,
         List<Field> declaredFields = Arrays.asList(awaitingInjectionClazz.getDeclaredFields());
         declaredFields = ListUtil.filter(declaredFields, notFinalOrStatic);
 
-        Collections.sort(declaredFields, superTypesLast);
+        sortSuperTypesLast(declaredFields);
 
         return declaredFields;
     }
 
-    static class FieldTypeAndNameComparator implements Comparator<Field> {
-        public int compare(Field field1, Field field2) {
-            Class<?> field1Type = field1.getType();
-            Class<?> field2Type = field2.getType();
+    /**
+     * Sort first by name, then move any fields after their supertypes.
+     */
+    static void sortSuperTypesLast(List<Field> fields) {
+        Collections.sort(fields, compareFieldsByName);
 
-            // if same type, compares on field name
-            if (field1Type == field2Type) {
-                return field1.getName().compareTo(field2.getName());
-            }
-            if(field1Type.isAssignableFrom(field2Type)) {
-                return 1;
+        int i = 0;
+
+        while (i < fields.size() - 1) {
+            Field f = fields.get(i);
+            Class<?> ft = f.getType();
+            int newPos = i;
+            for (int j = i + 1; j < fields.size(); j++) {
+                Class<?> t = fields.get(j).getType();
+
+                if (ft != t && ft.isAssignableFrom(t)) {
+                    newPos = j;
+                }
             }
-            if(field2Type.isAssignableFrom(field1Type)) {
-                return -1;
+
+            if (newPos == i) {
+                i++;
+            } else {
+                fields.remove(i);
+                fields.add(newPos, f);
```

Comment:
Also I didn't thought about that, but changing the type to a `TreeSet` may be interesting.